IDF Component not compatible with ESP IDF v4

[IDF component page states compatibility with ESP IDF >= 4.3](https://components.espressif.com/components/brianpugh/tamp/versions/1.10.0/dependencies?language=en)

### Environment
OS: Windows 11
ESP IDF: v4.4.1
Project: Cloned [demo repo](https://github.com/BrianPugh/esp-tamp-demo)

### Steps to Recreate
```
idf.py fullclean
idf.py build
```

### Results
```
[664/1044] Generating x509_crt_bundle
C:\Espressif\frameworks\esp-idf-v4.4.1\components\mbedtls\esp_crt_bundle\gen_crt_bundle.py:116: CryptographyDeprecationWarning: Parsed a serial number which wasn't positive (i.e., it was negative or zero), which is disallowed by RFC 5280. Loading this certificate will cause an exception in a future release of cryptography.
  self.certificates.append(x509.load_pem_x509_certificate(crt.encode(), default_backend()))
[1010/1044] Building CXX object esp-idf/brianpugh__tamp/CMakeFiles/__idf_brianpugh__tamp.dir/compressor_esp32.cpp.obj
FAILED: esp-idf/brianpugh__tamp/CMakeFiles/__idf_brianpugh__tamp.dir/compressor_esp32.cpp.obj
ccache C:\Espressif\tools\xtensa-esp32-elf\esp-2021r2-patch3-8.4.0\xtensa-esp32-elf\bin\xtensa-esp32-elf-g++.exe  -Iconfig -I../managed_components/brianpugh__tamp -I../managed_components/brianpugh__tamp/tamp -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/newlib/platform_include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/freertos/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/freertos/include/esp_additions/freertos -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/freertos/port/xtensa/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/freertos/include/esp_additions -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_hw_support/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_hw_support/include/soc -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_hw_support/include/soc/esp32 -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_hw_support/port/esp32/. -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_hw_support/port/esp32/private_include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/heap/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/log/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/lwip/include/apps -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/lwip/include/apps/sntp -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/lwip/lwip/src/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/lwip/port/esp32/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/lwip/port/esp32/include/arch -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/soc/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/soc/esp32/. -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/soc/esp32/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/hal/esp32/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/hal/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/hal/platform_port/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/include/esp32 -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/esp32 -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_common/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_system/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_system/port/soc -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_system/port/public_compat -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp32/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/xtensa/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/xtensa/esp32/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/driver/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/driver/esp32/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_pm/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_ringbuf/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/efuse/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/efuse/esp32/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/vfs/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_wifi/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_event/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_netif/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_eth/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/tcpip_adapter/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_phy/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_phy/esp32/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_ipc/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/app_trace/include -IC:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_timer/include -mlongcalls -Wno-frame-address -ffunction-sections -fdata-sections -Wall -Werror=all -Wno-error=unused-function -Wno-error=unused-variable -Wno-error=deprecated-declarations -Wextra -Wno-unused-parameter -Wno-sign-compare -ggdb -Og -fmacro-prefix-map=C:/Users/MattLemcke/development/esp-tamp-demo=. -fmacro-prefix-map=C:/Espressif/frameworks/esp-idf-v4.4.1=IDF -fstrict-volatile-bitfields -Wno-error=unused-but-set-variable -fno-jump-tables -fno-tree-switch-conversion -std=gnu++11 -fno-exceptions -fno-rtti -D_GNU_SOURCE -DIDF_VER=\"v4.4.1\" -DESP_PLATFORM -D_POSIX_READER_WRITER_LOCKS -std=gnu++2a -MD -MT esp-idf/brianpugh__tamp/CMakeFiles/__idf_brianpugh__tamp.dir/compressor_esp32.cpp.obj -MF esp-idf\brianpugh__tamp\CMakeFiles\__idf_brianpugh__tamp.dir\compressor_esp32.cpp.obj.d -o esp-idf/brianpugh__tamp/CMakeFiles/__idf_brianpugh__tamp.dir/compressor_esp32.cpp.obj -c ../managed_components/brianpugh__tamp/compressor_esp32.cpp
In file included from ../managed_components/brianpugh__tamp/compressor_esp32.cpp:16:
../managed_components/brianpugh__tamp/private/tamp_search.hpp:20:10: fatal error: span: No such file or directory
 #include <span>
          ^~~~~~
compilation terminated.
[1023/1044] Performing configure step for 'bootloader'
-- Found Git: C:/Users/MattLemcke/AppData/Local/Programs/Git/cmd/git.exe (found version "2.46.0.windows.1")
-- The C compiler identification is GNU 8.4.0
-- The CXX compiler identification is GNU 8.4.0
-- The ASM compiler identification is GNU
-- Found assembler: C:/Espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/xtensa-esp32-elf-gcc.exe     
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/xtensa-esp32-elf-gcc.exe - skipped
lf-g++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building ESP-IDF components for target esp32
-- Project sdkconfig file C:/Users/MattLemcke/development/esp-tamp-demo/sdkconfig
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/soc/esp32/ld/esp32.peripherals.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/esp32/ld/esp32.rom.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/esp32/ld/esp32.rom.api.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/esp32/ld/esp32.rom.libgcc.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/esp32/ld/esp32.rom.newlib-funcs.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader/subproject/main/ld/esp32/bootloader.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader/subproject/main/ld/esp32/bootloader.rom.ld
-- Components: bootloader bootloader_support efuse esp32 esp_common esp_hw_support esp_rom esp_system esptool_py freertos hal log main micro-ecc newlib partition_table soc spi_flash xtensa
-- Component paths: C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader_support C:/Espressif/frameworks/esp-idf-v4.4.1/components/efuse C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp32 C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_common C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_hw_support C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_system C:/Espressif/frameworks/esp-idf-v4.4.1/components/esptool_py C:/Espressif/frameworks/esp-idf-v4.4.1/components/freertos C:/Espressif/frameworks/esp-idf-v4.4.1/components/hal C:/Espressif/frameworks/esp-idf-v4.4.1/components/log C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader/subproject/main C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader/subproject/components/micro-ecc C:/Espressif/frameworks/esp-idf-v4.4.1/components/newlib C:/Espressif/frameworks/esp-idf-v4.4.1/components/partition_table C:/Espressif/frameworks/esp-idf-v4.4.1/components/soc C:/Espressif/frameworks/esp-idf-v4.4.1/components/spi_flash C:/Espressif/frameworks/esp-idf-v4.4.1/components/xtensa
```
Lines of particular interest:
```
In file included from ../managed_components/brianpugh__tamp/compressor_esp32.cpp:16:
../managed_components/brianpugh__tamp/private/tamp_search.hpp:20:10: fatal error: span: No such file or directory
 #include <span>
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -65,6 +65,7 @@ repos:
       - id: clang-format
         types_or: [c, c++]
         args: ['-style=file', '-i']
+        exclude: ^espidf/tamp/private/tamp_search\.hpp$
 
   - repo: https://github.com/pocc/pre-commit-hooks
     rev: v1.3.5

--- a/espidf/tamp/CMakeLists.txt
+++ b/espidf/tamp/CMakeLists.txt
@@ -7,7 +7,13 @@
 # CMakeLists in this exact order for cmake to work correctly
 cmake_minimum_required(VERSION 3.5)
 
-set(CMAKE_CXX_STANDARD 20)
+if(IDF_VER_MAJOR EQUAL 4 AND IDF_VER_MINOR LESS 5)
+    # ESP-IDF 4.0 -> 4.4: default compiler is C++11; enable C++17 safely
+    set(CMAKE_CXX_STANDARD 17)
+else()
+    # ESP-IDF >= 4.5 or 5.x: use C++20
+    set(CMAKE_CXX_STANDARD 20)
+endif()
 set(CMAKE_CXX_STANDARD_REQUIRED True)
 
 

--- a/espidf/tamp/private/copyutil.hpp
+++ b/espidf/tamp/private/copyutil.hpp
@@ -90,8 +90,8 @@ static constexpr uint32_t MSB = (1 << log2<N>());
  * @param len
  */
 template <uint32_t CEIL = 16>
-    requires(CEIL != 0 /* && (CEIL & (CEIL-1)) == 0 */)
 static inline void __attribute__((always_inline)) unrolled_cpy(void* dst, const void* src, std::size_t len) noexcept {
+    static_assert(CEIL != 0 /* && (CEIL & (CEIL-1)) == 0 */);
     if constexpr (CEIL > 32) {
         // You should probably be using std::memcpy...
         {
@@ -120,6 +120,10 @@ static inline void __attribute__((always_inline)) unrolled_cpy(void* dst, const 
             incptr<CEIL / 2>(src);
         }
         unrolled_cpy<CEIL / 2>(dst, src, len);
+    } else {
+        (void)dst;
+        (void)src;
+        (void)len;
     }
 }
 }  // namespace

--- a/espidf/tamp/private/tamp_search.hpp
+++ b/espidf/tamp/private/tamp_search.hpp
@@ -354,41 +354,41 @@ class Locator {
                 if constexpr (Arch::XTENSA && Arch::XT_LOOP) {
                     uint32_t tmp;
 
-                        asm("LOOPNEZ %[len], end_%="
-                            "\n"
-                            "L32I %[tmp], %[data], 0"
-                            "\n"
-                            "ADDI %[data], %[data], 1"
-                            "\n"  // Pipelining.
-                            "BEQ %[tmp], %[v], found_%="
-                            "\n"
-                            "end_%=:"
-                            "\n"
-                            "MOVI %[data], 1"
-                            "\n"  // Make result = 0 w/o a jump
-                            "found_%=:"
-                            "\n"
-                            "ADDI %[data], %[data], -1"
-                            "\n"
-                            "exit_%=:"
-                            : [tmp] "=&r"(tmp), [data] "+&r"(data)
-                            : [v] "r"(v), [len] "r"(dataLen - (sizeof(uint32_t) - 1)),
-                              "m"(*(const uint8_t(*)[dataLen])data));
+                    asm("LOOPNEZ %[len], end_%="
+                        "\n"
+                        "L32I %[tmp], %[data], 0"
+                        "\n"
+                        "ADDI %[data], %[data], 1"
+                        "\n"  // Pipelining.
+                        "BEQ %[tmp], %[v], found_%="
+                        "\n"
+                        "end_%=:"
+                        "\n"
+                        "MOVI %[data], 1"
+                        "\n"  // Make result = 0 w/o a jump
+                        "found_%=:"
+                        "\n"
+                        "ADDI %[data], %[data], -1"
+                        "\n"
+                        "exit_%=:"
+                        : [tmp] "=&r"(tmp), [data] "+&r"(data)
+                        : [v] "r"(v), [len] "r"(dataLen - (sizeof(uint32_t) - 1)),
+                          "m"(*(const uint8_t(*)[dataLen])data));
 
-                        return data;
+                    return data;
 
-                    } else {
-                        constexpr uint32_t LOOP_UNROLL_FACTOR = 8;
+                } else {
+                    constexpr uint32_t LOOP_UNROLL_FACTOR = 8;
 
-                        const uint8_t* const end_unrolled = data + multof<LOOP_UNROLL_FACTOR>(dataLen - (patLen - 1));
-                        while (data < end_unrolled && !unrolled_find<uint32_t, LOOP_UNROLL_FACTOR>(data, v)) TAMP_CPP_LIKELY {
-                            incptr<LOOP_UNROLL_FACTOR>(data);
-                        }
+                    const uint8_t* const end_unrolled = data + multof<LOOP_UNROLL_FACTOR>(dataLen - (patLen - 1));
+                    while (data < end_unrolled && !unrolled_find<uint32_t, LOOP_UNROLL_FACTOR>(data, v)) TAMP_CPP_LIKELY {
+                        incptr<LOOP_UNROLL_FACTOR>(data);
+                    }
 
-                        if (data >= end_unrolled) {
-                            // Nothing found so far.
-                            while (data < end && as<uint32_t>(data) != v) {
-                                incptr<1>(data);
+                    if (data >= end_unrolled) {
+                        // Nothing found so far.
+                        while (data < end && as<uint32_t>(data) != v) {
+                            incptr<1>(data);
                         }
                     }
                 }
@@ -526,15 +526,15 @@ class Locator {
                 if constexpr (Arch::XTENSA && Arch::XT_LOOP) {
                     uint32_t tmp = dataLen;
                     asm volatile(
-                            "LOOPNEZ %[tmp], end_%="
-                            "\n"
-                            "L8UI %[tmp], %[data], 0"
-                            "\n"
-                            "BEQ %[tmp], %[v], end_%="
-                            "\n"
-                            "ADDI %[data], %[data], 1"
-                            "\n"
-                            "end_%=:"
+                        "LOOPNEZ %[tmp], end_%="
+                        "\n"
+                        "L8UI %[tmp], %[data], 0"
+                        "\n"
+                        "BEQ %[tmp], %[v], end_%="
+                        "\n"
+                        "ADDI %[data], %[data], 1"
+                        "\n"
+                        "end_%=:"
                         : [tmp] "+r"(tmp), [data] "+r"(data)
                         : [v] "r"(v));
                 } else {

--- a/espidf/tamp/private/tamp_search.hpp
+++ b/espidf/tamp/private/tamp_search.hpp
@@ -29,11 +29,9 @@
 #include "tamp_arch.hpp"
 
 #if __cplusplus >= 202002L
-// #define TAMP_REQUIRES(...) requires(__VA_ARGS__)
 #define TAMP_CPP_LIKELY [[likely]]
 #define TAMP_CPP_UNLIKELY [[unlikely]]
 #else
-// #define TAMP_REQUIRES(...) /* nothing */
 #define TAMP_CPP_LIKELY
 #define TAMP_CPP_UNLIKELY
 #endif

--- a/espidf/tamp/private/tamp_search.hpp
+++ b/espidf/tamp/private/tamp_search.hpp
@@ -285,13 +285,12 @@ class Locator {
     template <typename T, uint32_t M, uint32_t N = 0>
     static bool __attribute__((always_inline)) unrolled_find(const uint8_t*& data, const uint32_t v) noexcept {
         if (as<T>(data + N) != v) TAMP_CPP_LIKELY {
-                if constexpr (N + 1 < M) {
-                    return unrolled_find<T, M, N + 1>(data, v);
-                } else {
-                    return false;
-                }
+            if constexpr (N + 1 < M) {
+                return unrolled_find<T, M, N + 1>(data, v);
+            } else {
+                return false;
             }
-        else {
+        } else {
             incptr<N>(data);
             return true;
         }
@@ -309,25 +308,23 @@ class Locator {
         static_assert(N < M);
         if constexpr (N % 2 == 0) {
             if ((as<uint32_t>(data + ((N / 2) * 2)) << 8) != vl) TAMP_CPP_LIKELY {
-                    if constexpr (N + 1 < M) {
-                        return unrolled_find_3<M, N + 1>(data, vl, vh);
-                    } else {
-                        return false;
-                    }
+                if constexpr (N + 1 < M) {
+                    return unrolled_find_3<M, N + 1>(data, vl, vh);
+                } else {
+                    return false;
                 }
-            else {
+            } else {
                 incptr<N>(data);
                 return true;
             }
         } else {
             if ((as<uint32_t>(data + ((N / 2) * 2)) >> 8) != vh) TAMP_CPP_LIKELY {
-                    if constexpr (N + 1 < M) {
-                        return unrolled_find_3<M, N + 1>(data, vl, vh);
-                    } else {
-                        return false;
-                    }
+                if constexpr (N + 1 < M) {
+                    return unrolled_find_3<M, N + 1>(data, vl, vh);
+                } else {
+                    return false;
                 }
-            else {
+            } else {
                 incptr<N>(data);
                 return true;
             }
@@ -352,10 +349,10 @@ class Locator {
 
         if (patLen > sizeof(uint16_t)) {
             if (patLen == sizeof(uint32_t)) TAMP_CPP_UNLIKELY {
-                    const uint32_t v = as<uint32_t>(pattern);
+                const uint32_t v = as<uint32_t>(pattern);
 
-                    if constexpr (Arch::XTENSA && Arch::XT_LOOP) {
-                        uint32_t tmp;
+                if constexpr (Arch::XTENSA && Arch::XT_LOOP) {
+                    uint32_t tmp;
 
                         asm("LOOPNEZ %[len], end_%="
                             "\n"
@@ -384,18 +381,18 @@ class Locator {
                         constexpr uint32_t LOOP_UNROLL_FACTOR = 8;
 
                         const uint8_t* const end_unrolled = data + multof<LOOP_UNROLL_FACTOR>(dataLen - (patLen - 1));
-                        while (data < end_unrolled && !unrolled_find<uint32_t, LOOP_UNROLL_FACTOR>(data, v))
-                            TAMP_CPP_LIKELY { incptr<LOOP_UNROLL_FACTOR>(data); }
+                        while (data < end_unrolled && !unrolled_find<uint32_t, LOOP_UNROLL_FACTOR>(data, v)) TAMP_CPP_LIKELY {
+                            incptr<LOOP_UNROLL_FACTOR>(data);
+                        }
 
                         if (data >= end_unrolled) {
                             // Nothing found so far.
                             while (data < end && as<uint32_t>(data) != v) {
                                 incptr<1>(data);
-                            }
                         }
                     }
                 }
-            else {
+            } else {
                 // assert patLen == 3
 
                 // Given a 24-bit value in a 32-bit variable V and 4 bytes (32 bits) loaded from memory
@@ -457,8 +454,8 @@ class Locator {
                     const uint8_t* const end_unrolled = data + multof<LOOP_UNROLL_FACTOR>(dataLen - (patLen - 1));
 
                     while (data < end_unrolled && !unrolled_find_3<LOOP_UNROLL_FACTOR>(data, vl, vh)) TAMP_CPP_LIKELY {
-                            incptr<LOOP_UNROLL_FACTOR>(data);
-                        }
+                        incptr<LOOP_UNROLL_FACTOR>(data);
+                    }
 
                     if (data >= end_unrolled) {
                         // Nothing found so far.
@@ -511,8 +508,9 @@ class Locator {
                     // Loop unrolled 8x.
                     const uint8_t* const end_unrolled = data + multof<LOOP_UNROLL_FACTOR>(dataLen - (patLen - 1));
 
-                    while (data < end_unrolled && !unrolled_find<uint16_t, LOOP_UNROLL_FACTOR>(data, v))
-                        TAMP_CPP_LIKELY { incptr<LOOP_UNROLL_FACTOR>(data); }
+                    while (data < end_unrolled && !unrolled_find<uint16_t, LOOP_UNROLL_FACTOR>(data, v)) TAMP_CPP_LIKELY {
+                        incptr<LOOP_UNROLL_FACTOR>(data);
+                    }
 
                     if (data >= end_unrolled) {
                         // Nothing found so far.
@@ -522,13 +520,12 @@ class Locator {
                     }
                 }
 
-            } else
-                TAMP_CPP_UNLIKELY {
-                    // assert patLen == 1
-                    const uint32_t v = as<uint8_t>(pattern);
-                    if constexpr (Arch::XTENSA && Arch::XT_LOOP) {
-                        uint32_t tmp = dataLen;
-                        asm volatile(
+            } else TAMP_CPP_UNLIKELY {
+                // assert patLen == 1
+                const uint32_t v = as<uint8_t>(pattern);
+                if constexpr (Arch::XTENSA && Arch::XT_LOOP) {
+                    uint32_t tmp = dataLen;
+                    asm volatile(
                             "LOOPNEZ %[tmp], end_%="
                             "\n"
                             "L8UI %[tmp], %[data], 0"
@@ -538,14 +535,14 @@ class Locator {
                             "ADDI %[data], %[data], 1"
                             "\n"
                             "end_%=:"
-                            : [tmp] "+r"(tmp), [data] "+r"(data)
-                            : [v] "r"(v));
-                    } else {
-                        while (data < end && as<uint8_t>(data) != v) {
-                            incptr<1>(data);
-                        }
+                        : [tmp] "+r"(tmp), [data] "+r"(data)
+                        : [v] "r"(v));
+                } else {
+                    while (data < end && as<uint8_t>(data) != v) {
+                        incptr<1>(data);
                     }
                 }
+            }
         }
         return (data < end) ? data : nullptr;
     }
@@ -555,13 +552,12 @@ class Locator {
     static bool __attribute__((always_inline))
     unrolled_find_f_l(const uint8_t*& first, const uint8_t*& last, const uint32_t f, const uint32_t l) noexcept {
         if (f != as<uint32_t>(first + N) || l != as<uint32_t>(last + N)) TAMP_CPP_LIKELY {
-                if constexpr (N + 1 < M) {
-                    return unrolled_find_f_l<M, N + 1>(first, last, f, l);
-                } else {
-                    return false;
-                }
+            if constexpr (N + 1 < M) {
+                return unrolled_find_f_l<M, N + 1>(first, last, f, l);
+            } else {
+                return false;
             }
-        else {
+        } else {
             incptr<N>(first);
             incptr<N>(last);
             return true;
@@ -578,9 +574,8 @@ class Locator {
             const uint8_t* ff = find_pattern_short_scalar((const uint8_t*)&f, sizeof(uint32_t), fp, end - fp);
             if (ff) {
                 if (*(const uint32_t*)(ff + len) != l) TAMP_CPP_LIKELY {
-                        fp = ff + 1;
-                    }
-                else {
+                    fp = ff + 1;
+                } else {
                     first = ff;
                     last = ff + len;
                     return true;
@@ -637,11 +632,10 @@ class Locator {
             {
                 constexpr uint32_t LOOP_UNROLL_FACTOR = 8;
                 const uint8_t* const end_unrolled = first + multof<LOOP_UNROLL_FACTOR>(end - first);
-                while (first < end_unrolled && !unrolled_find_f_l<LOOP_UNROLL_FACTOR>(first, last, f, l))
-                    TAMP_CPP_LIKELY {
-                        incptr<LOOP_UNROLL_FACTOR>(first);
-                        incptr<LOOP_UNROLL_FACTOR>(last);
-                    }
+                while (first < end_unrolled && !unrolled_find_f_l<LOOP_UNROLL_FACTOR>(first, last, f, l)) TAMP_CPP_LIKELY {
+                    incptr<LOOP_UNROLL_FACTOR>(first);
+                    incptr<LOOP_UNROLL_FACTOR>(last);
+                }
 
                 if (first >= end_unrolled) {
                     // Nothing found so far.
@@ -672,13 +666,12 @@ class Locator {
     template <uint32_t M, uint32_t N = 0, typename P>
     static bool __attribute__((always_inline)) unrolled_cmp8(const P*& d1, const P*& d2) noexcept {
         if (*(p<uint8_t>(d1) + N) == *(p<uint8_t>(d2) + N)) TAMP_CPP_LIKELY {
-                if constexpr (N + 1 < M) {
-                    return unrolled_cmp8<M, N + 1>(d1, d2);
-                } else {
-                    return true;
-                }
+            if constexpr (N + 1 < M) {
+                return unrolled_cmp8<M, N + 1>(d1, d2);
+            } else {
+                return true;
             }
-        else {
+        } else {
             incptr<N>(d1);
             incptr<N>(d2);
             return false;
@@ -735,10 +728,9 @@ class Locator {
 
             while (d1 < end) {
                 if (as<uint8_t>(d1) == as<uint8_t>(d2)) TAMP_CPP_LIKELY {
-                        incptr<1>(d1);
-                        incptr<1>(d2);
-                    }
-                else {
+                    incptr<1>(d1);
+                    incptr<1>(d2);
+                } else {
                     break;
                 }
             }
@@ -1044,12 +1036,11 @@ class Locator {
                             tmp = tmp << bits;  // Remove the bit we're handling now.
                         }
                         if (s1 <= end) TAMP_CPP_LIKELY {
-                                if (cmpLen == 0 || cmp8(s1, pat1, cmpLen) >= cmpLen) {
-                                    // stats.matchFound(patLen, (s1-1)-data + patLen);
-                                    return s1 - 1;
-                                }
+                            if (cmpLen == 0 || cmp8(s1, pat1, cmpLen) >= cmpLen) {
+                                // stats.matchFound(patLen, (s1-1)-data + patLen);
+                                return s1 - 1;
                             }
-                        else {
+                        } else {
                             // Found match would extend beyond the end of data.
                             return nullptr;
                         }
@@ -1089,10 +1080,10 @@ class Locator {
                     bestMatch = match;
                     matchLen = searchLen;
                     if (searchLen < patLen) TAMP_CPP_LIKELY {
-                            // If the current match happens to extend beyond what we searched for,
-                            // we'll take that too.
-                            matchLen += cmp8(pattern + searchLen, match + searchLen, patLen - searchLen);
-                        }
+                        // If the current match happens to extend beyond what we searched for,
+                        // we'll take that too.
+                        matchLen += cmp8(pattern + searchLen, match + searchLen, patLen - searchLen);
+                    }
                     dataLen -= match + 1 - data;
                     data = match + 1;
 

--- a/espidf/tamp/private/tamp_search.hpp
+++ b/espidf/tamp/private/tamp_search.hpp
@@ -20,13 +20,23 @@
 
 #include <cstdint>
 
-#if __cplusplus > 201703L
+#if __has_include(<span>) && __cplusplus > 201703L
 #include <span>
 #endif
 #include <algorithm>
 #include <type_traits>
 
 #include "tamp_arch.hpp"
+
+#if __cplusplus >= 202002L
+// #define TAMP_REQUIRES(...) requires(__VA_ARGS__)
+#define TAMP_CPP_LIKELY [[likely]]
+#define TAMP_CPP_UNLIKELY [[unlikely]]
+#else
+// #define TAMP_REQUIRES(...) /* nothing */
+#define TAMP_CPP_LIKELY
+#define TAMP_CPP_UNLIKELY
+#endif
 
 namespace tamp {
 
@@ -274,13 +284,14 @@ class Locator {
 
     template <typename T, uint32_t M, uint32_t N = 0>
     static bool __attribute__((always_inline)) unrolled_find(const uint8_t*& data, const uint32_t v) noexcept {
-        if (as<T>(data + N) != v) [[likely]] {
-            if constexpr (N + 1 < M) {
-                return unrolled_find<T, M, N + 1>(data, v);
-            } else {
-                return false;
+        if (as<T>(data + N) != v) TAMP_CPP_LIKELY {
+                if constexpr (N + 1 < M) {
+                    return unrolled_find<T, M, N + 1>(data, v);
+                } else {
+                    return false;
+                }
             }
-        } else {
+        else {
             incptr<N>(data);
             return true;
         }
@@ -297,24 +308,26 @@ class Locator {
          */
         static_assert(N < M);
         if constexpr (N % 2 == 0) {
-            if ((as<uint32_t>(data + ((N / 2) * 2)) << 8) != vl) [[likely]] {
-                if constexpr (N + 1 < M) {
-                    return unrolled_find_3<M, N + 1>(data, vl, vh);
-                } else {
-                    return false;
+            if ((as<uint32_t>(data + ((N / 2) * 2)) << 8) != vl) TAMP_CPP_LIKELY {
+                    if constexpr (N + 1 < M) {
+                        return unrolled_find_3<M, N + 1>(data, vl, vh);
+                    } else {
+                        return false;
+                    }
                 }
-            } else {
+            else {
                 incptr<N>(data);
                 return true;
             }
         } else {
-            if ((as<uint32_t>(data + ((N / 2) * 2)) >> 8) != vh) [[likely]] {
-                if constexpr (N + 1 < M) {
-                    return unrolled_find_3<M, N + 1>(data, vl, vh);
-                } else {
-                    return false;
+            if ((as<uint32_t>(data + ((N / 2) * 2)) >> 8) != vh) TAMP_CPP_LIKELY {
+                    if constexpr (N + 1 < M) {
+                        return unrolled_find_3<M, N + 1>(data, vl, vh);
+                    } else {
+                        return false;
+                    }
                 }
-            } else {
+            else {
                 incptr<N>(data);
                 return true;
             }
@@ -338,51 +351,51 @@ class Locator {
         const uint8_t* const end = data + dataLen - (patLen - 1);
 
         if (patLen > sizeof(uint16_t)) {
-            if (patLen == sizeof(uint32_t)) [[unlikely]] {
-                const uint32_t v = as<uint32_t>(pattern);
+            if (patLen == sizeof(uint32_t)) TAMP_CPP_UNLIKELY {
+                    const uint32_t v = as<uint32_t>(pattern);
 
-                if constexpr (Arch::XTENSA && Arch::XT_LOOP) {
-                    uint32_t tmp;
+                    if constexpr (Arch::XTENSA && Arch::XT_LOOP) {
+                        uint32_t tmp;
 
-                    asm("LOOPNEZ %[len], end_%="
-                        "\n"
-                        "L32I %[tmp], %[data], 0"
-                        "\n"
-                        "ADDI %[data], %[data], 1"
-                        "\n"  // Pipelining.
-                        "BEQ %[tmp], %[v], found_%="
-                        "\n"
-                        "end_%=:"
-                        "\n"
-                        "MOVI %[data], 1"
-                        "\n"  // Make result = 0 w/o a jump
-                        "found_%=:"
-                        "\n"
-                        "ADDI %[data], %[data], -1"
-                        "\n"
-                        "exit_%=:"
-                        : [tmp] "=&r"(tmp), [data] "+&r"(data)
-                        : [v] "r"(v), [len] "r"(dataLen - (sizeof(uint32_t) - 1)),
-                          "m"(*(const uint8_t(*)[dataLen])data));
+                        asm("LOOPNEZ %[len], end_%="
+                            "\n"
+                            "L32I %[tmp], %[data], 0"
+                            "\n"
+                            "ADDI %[data], %[data], 1"
+                            "\n"  // Pipelining.
+                            "BEQ %[tmp], %[v], found_%="
+                            "\n"
+                            "end_%=:"
+                            "\n"
+                            "MOVI %[data], 1"
+                            "\n"  // Make result = 0 w/o a jump
+                            "found_%=:"
+                            "\n"
+                            "ADDI %[data], %[data], -1"
+                            "\n"
+                            "exit_%=:"
+                            : [tmp] "=&r"(tmp), [data] "+&r"(data)
+                            : [v] "r"(v), [len] "r"(dataLen - (sizeof(uint32_t) - 1)),
+                              "m"(*(const uint8_t(*)[dataLen])data));
 
-                    return data;
+                        return data;
 
-                } else {
-                    constexpr uint32_t LOOP_UNROLL_FACTOR = 8;
+                    } else {
+                        constexpr uint32_t LOOP_UNROLL_FACTOR = 8;
 
-                    const uint8_t* const end_unrolled = data + multof<LOOP_UNROLL_FACTOR>(dataLen - (patLen - 1));
-                    while (data < end_unrolled && !unrolled_find<uint32_t, LOOP_UNROLL_FACTOR>(data, v)) [[likely]] {
-                        incptr<LOOP_UNROLL_FACTOR>(data);
-                    }
+                        const uint8_t* const end_unrolled = data + multof<LOOP_UNROLL_FACTOR>(dataLen - (patLen - 1));
+                        while (data < end_unrolled && !unrolled_find<uint32_t, LOOP_UNROLL_FACTOR>(data, v))
+                            TAMP_CPP_LIKELY { incptr<LOOP_UNROLL_FACTOR>(data); }
 
-                    if (data >= end_unrolled) {
-                        // Nothing found so far.
-                        while (data < end && as<uint32_t>(data) != v) {
-                            incptr<1>(data);
+                        if (data >= end_unrolled) {
+                            // Nothing found so far.
+                            while (data < end && as<uint32_t>(data) != v) {
+                                incptr<1>(data);
+                            }
                         }
                     }
                 }
-            } else {
+            else {
                 // assert patLen == 3
 
                 // Given a 24-bit value in a 32-bit variable V and 4 bytes (32 bits) loaded from memory
@@ -443,9 +456,9 @@ class Locator {
                     // Loop unrolled 8x.
                     const uint8_t* const end_unrolled = data + multof<LOOP_UNROLL_FACTOR>(dataLen - (patLen - 1));
 
-                    while (data < end_unrolled && !unrolled_find_3<LOOP_UNROLL_FACTOR>(data, vl, vh)) [[likely]] {
-                        incptr<LOOP_UNROLL_FACTOR>(data);
-                    }
+                    while (data < end_unrolled && !unrolled_find_3<LOOP_UNROLL_FACTOR>(data, vl, vh)) TAMP_CPP_LIKELY {
+                            incptr<LOOP_UNROLL_FACTOR>(data);
+                        }
 
                     if (data >= end_unrolled) {
                         // Nothing found so far.
@@ -498,9 +511,8 @@ class Locator {
                     // Loop unrolled 8x.
                     const uint8_t* const end_unrolled = data + multof<LOOP_UNROLL_FACTOR>(dataLen - (patLen - 1));
 
-                    while (data < end_unrolled && !unrolled_find<uint16_t, LOOP_UNROLL_FACTOR>(data, v)) [[likely]] {
-                        incptr<LOOP_UNROLL_FACTOR>(data);
-                    }
+                    while (data < end_unrolled && !unrolled_find<uint16_t, LOOP_UNROLL_FACTOR>(data, v))
+                        TAMP_CPP_LIKELY { incptr<LOOP_UNROLL_FACTOR>(data); }
 
                     if (data >= end_unrolled) {
                         // Nothing found so far.
@@ -510,29 +522,30 @@ class Locator {
                     }
                 }
 
-            } else [[unlikely]] {
-                // assert patLen == 1
-                const uint32_t v = as<uint8_t>(pattern);
-                if constexpr (Arch::XTENSA && Arch::XT_LOOP) {
-                    uint32_t tmp = dataLen;
-                    asm volatile(
-                        "LOOPNEZ %[tmp], end_%="
-                        "\n"
-                        "L8UI %[tmp], %[data], 0"
-                        "\n"
-                        "BEQ %[tmp], %[v], end_%="
-                        "\n"
-                        "ADDI %[data], %[data], 1"
-                        "\n"
-                        "end_%=:"
-                        : [tmp] "+r"(tmp), [data] "+r"(data)
-                        : [v] "r"(v));
-                } else {
-                    while (data < end && as<uint8_t>(data) != v) {
-                        incptr<1>(data);
+            } else
+                TAMP_CPP_UNLIKELY {
+                    // assert patLen == 1
+                    const uint32_t v = as<uint8_t>(pattern);
+                    if constexpr (Arch::XTENSA && Arch::XT_LOOP) {
+                        uint32_t tmp = dataLen;
+                        asm volatile(
+                            "LOOPNEZ %[tmp], end_%="
+                            "\n"
+                            "L8UI %[tmp], %[data], 0"
+                            "\n"
+                            "BEQ %[tmp], %[v], end_%="
+                            "\n"
+                            "ADDI %[data], %[data], 1"
+                            "\n"
+                            "end_%=:"
+                            : [tmp] "+r"(tmp), [data] "+r"(data)
+                            : [v] "r"(v));
+                    } else {
+                        while (data < end && as<uint8_t>(data) != v) {
+                            incptr<1>(data);
+                        }
                     }
                 }
-            }
         }
         return (data < end) ? data : nullptr;
     }
@@ -541,13 +554,14 @@ class Locator {
     template <uint32_t M, uint32_t N = 0>
     static bool __attribute__((always_inline))
     unrolled_find_f_l(const uint8_t*& first, const uint8_t*& last, const uint32_t f, const uint32_t l) noexcept {
-        if (f != as<uint32_t>(first + N) || l != as<uint32_t>(last + N)) [[likely]] {
-            if constexpr (N + 1 < M) {
-                return unrolled_find_f_l<M, N + 1>(first, last, f, l);
-            } else {
-                return false;
+        if (f != as<uint32_t>(first + N) || l != as<uint32_t>(last + N)) TAMP_CPP_LIKELY {
+                if constexpr (N + 1 < M) {
+                    return unrolled_find_f_l<M, N + 1>(first, last, f, l);
+                } else {
+                    return false;
+                }
             }
-        } else {
+        else {
             incptr<N>(first);
             incptr<N>(last);
             return true;
@@ -563,9 +577,10 @@ class Locator {
         do {
             const uint8_t* ff = find_pattern_short_scalar((const uint8_t*)&f, sizeof(uint32_t), fp, end - fp);
             if (ff) {
-                if (*(const uint32_t*)(ff + len) != l) [[likely]] {
-                    fp = ff + 1;
-                } else {
+                if (*(const uint32_t*)(ff + len) != l) TAMP_CPP_LIKELY {
+                        fp = ff + 1;
+                    }
+                else {
                     first = ff;
                     last = ff + len;
                     return true;
@@ -578,7 +593,7 @@ class Locator {
         // do {
         //     uint32_t ftmp1 = *(const uint32_t*)(fp++);
         //     uint32_t ftmp2 = *(const uint32_t*)(lp++);
-        //     if(f == ftmp1 && l == ftmp2) [[unlikely]] {
+        //     if(f == ftmp1 && l == ftmp2) TAMP_CPP_UNLIKELY {
         //         first = fp-1;
         //         last = lp-1;
         //         break;
@@ -622,10 +637,11 @@ class Locator {
             {
                 constexpr uint32_t LOOP_UNROLL_FACTOR = 8;
                 const uint8_t* const end_unrolled = first + multof<LOOP_UNROLL_FACTOR>(end - first);
-                while (first < end_unrolled && !unrolled_find_f_l<LOOP_UNROLL_FACTOR>(first, last, f, l)) [[likely]] {
-                    incptr<LOOP_UNROLL_FACTOR>(first);
-                    incptr<LOOP_UNROLL_FACTOR>(last);
-                }
+                while (first < end_unrolled && !unrolled_find_f_l<LOOP_UNROLL_FACTOR>(first, last, f, l))
+                    TAMP_CPP_LIKELY {
+                        incptr<LOOP_UNROLL_FACTOR>(first);
+                        incptr<LOOP_UNROLL_FACTOR>(last);
+                    }
 
                 if (first >= end_unrolled) {
                     // Nothing found so far.
@@ -655,13 +671,14 @@ class Locator {
 
     template <uint32_t M, uint32_t N = 0, typename P>
     static bool __attribute__((always_inline)) unrolled_cmp8(const P*& d1, const P*& d2) noexcept {
-        if (*(p<uint8_t>(d1) + N) == *(p<uint8_t>(d2) + N)) [[likely]] {
-            if constexpr (N + 1 < M) {
-                return unrolled_cmp8<M, N + 1>(d1, d2);
-            } else {
-                return true;
+        if (*(p<uint8_t>(d1) + N) == *(p<uint8_t>(d2) + N)) TAMP_CPP_LIKELY {
+                if constexpr (N + 1 < M) {
+                    return unrolled_cmp8<M, N + 1>(d1, d2);
+                } else {
+                    return true;
+                }
             }
-        } else {
+        else {
             incptr<N>(d1);
             incptr<N>(d2);
             return false;
@@ -717,10 +734,11 @@ class Locator {
             const uint8_t* const end = p<uint8_t>(d1) + len;
 
             while (d1 < end) {
-                if (as<uint8_t>(d1) == as<uint8_t>(d2)) [[likely]] {
-                    incptr<1>(d1);
-                    incptr<1>(d2);
-                } else {
+                if (as<uint8_t>(d1) == as<uint8_t>(d2)) TAMP_CPP_LIKELY {
+                        incptr<1>(d1);
+                        incptr<1>(d2);
+                    }
+                else {
                     break;
                 }
             }
@@ -1025,12 +1043,13 @@ class Locator {
                             s1 += bits;
                             tmp = tmp << bits;  // Remove the bit we're handling now.
                         }
-                        if (s1 <= end) [[likely]] {
-                            if (cmpLen == 0 || cmp8(s1, pat1, cmpLen) >= cmpLen) {
-                                // stats.matchFound(patLen, (s1-1)-data + patLen);
-                                return s1 - 1;
+                        if (s1 <= end) TAMP_CPP_LIKELY {
+                                if (cmpLen == 0 || cmp8(s1, pat1, cmpLen) >= cmpLen) {
+                                    // stats.matchFound(patLen, (s1-1)-data + patLen);
+                                    return s1 - 1;
+                                }
                             }
-                        } else {
+                        else {
                             // Found match would extend beyond the end of data.
                             return nullptr;
                         }
@@ -1069,11 +1088,11 @@ class Locator {
                 if (match) {
                     bestMatch = match;
                     matchLen = searchLen;
-                    if (searchLen < patLen) [[likely]] {
-                        // If the current match happens to extend beyond what we searched for,
-                        // we'll take that too.
-                        matchLen += cmp8(pattern + searchLen, match + searchLen, patLen - searchLen);
-                    }
+                    if (searchLen < patLen) TAMP_CPP_LIKELY {
+                            // If the current match happens to extend beyond what we searched for,
+                            // we'll take that too.
+                            matchLen += cmp8(pattern + searchLen, match + searchLen, patLen - searchLen);
+                        }
                     dataLen -= match + 1 - data;
                     data = match + 1;
 

